pkg/cloudprovider/aws/controllers/infrastructure/controller.go

```diff
@@ -0,0 +1,343 @@
+		Name:      injection.GetOptions(ctx).DeploymentName,
+		Namespace: system.Namespace(),
+	}
+
+	notFound := false
+	if err := retry.Do(func() error {
+		err := c.kubeClient.Get(ctx, nn, dep)
+		if apierrors.IsNotFound(err) {
+			notFound = true
+		}
+		return client.IgnoreNotFound(err)
+	}); err != nil {
+		logging.FromContext(ctx).Errorf("Getting the deployment %s for cleanup, %v", nn, err)
+	}
+
+	// Deployment is already deleted or currently deleting, so we should cleanup the infrastructure
+	if notFound || !dep.DeletionTimestamp.IsZero() {
+		if err := retry.Do(func() error { return c.DeleteInfrastructure(ctx) }); err != nil {
+			logging.FromContext(ctx).Errorf("Deprovisioning the infrastructure, %v", err)
+		}
+	}
+}
+
+// Ready returns a channel that serves as a gate for other controllers
+// to wait on the infrastructure to be in a good state. When the infrastructure is ready,
+// this channel is closed so other controllers can proceed with their operations
+func (c *Controller) Ready() <-chan struct{} {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.readinessChan
+}
+
+func (c *Controller) Trigger() {
+	c.trigger <- struct{}{}
+}
+
+func (c *Controller) Done() <-chan struct{} {
+	return c.done
+}
+
+func (c *Controller) setReady(ctx context.Context, ready bool) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// If the infrastructure we close the readiness channel to let all
+	// other channels that are waiting on Ready() proceed; otherwise, open
+	// a channel to tell the other goroutines to wait
+	if ready {
+		healthy.Set(1)
+		if c.ready != ready {
+			logging.FromContext(ctx).Infof("Infrastructure is healthy")
+			c.recorder.InfrastructureHealthy(ctx, c.kubeClient)
+			close(c.readinessChan)
+		}
+	} else {
+		healthy.Set(0)
+		if c.ready != ready {
+			logging.FromContext(ctx).Infof("Infrastructure is unhealthy")
+			c.recorder.InfrastructureUnhealthy(ctx, c.kubeClient)
+		}
+		c.readinessChan = make(chan struct{})
+	}
+	c.ready = ready
+	c.ready = ready
+}
+
+// EnsureInfrastructure reconciles the SQS queue and the EventBridge rules with the expected
+// configuration prescribed by Karpenter
+func (c *Controller) EnsureInfrastructure(ctx context.Context) (err error) {
+	defer metrics.Measure(reconcileDuration)()
+
+	wg := &sync.WaitGroup{}
+	m := &sync.Mutex{}
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		e := c.ensureQueue(ctx)
+		m.Lock()
+		err = multierr.Append(err, e)
+		m.Unlock()
+	}()
+	go func() {
+		defer wg.Done()
+		e := c.ensureEventBridge(ctx)
+		m.Lock()
+		err = multierr.Append(err, e)
+		m.Unlock()
+	}()
+	wg.Wait()
+	return err
+}
+
+// DeleteInfrastructure removes the infrastructure that was stood up and reconciled
+// by the infrastructure controller for SQS message polling
+func (c *Controller) DeleteInfrastructure(ctx context.Context) (err error) {
+	logging.FromContext(ctx).Infof("Deprovisioning the infrastructure...")
+	wg := &sync.WaitGroup{}
+	m := &sync.Mutex{}
+
+	wg.Add(2)
+	go func() {
```

Comment:
Primarily because we are performing an asynchronous action (i.e. calling out to an API) and these calls don't rely on each other, so why not make it quicker?